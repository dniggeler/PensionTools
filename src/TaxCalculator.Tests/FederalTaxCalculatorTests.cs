using System;
using System.Threading.Tasks;
using Domain.Enums;
using FluentAssertions;
using PensionCoach.Tools.TaxCalculator.Abstractions;
using PensionCoach.Tools.TaxCalculator.Abstractions.Models.Person;
using Snapshooter.Xunit;
using Xunit;

namespace TaxCalculator.Tests
{
    [Trait("Calculator", "Federal")]
    public class FederalTaxCalculatorTests : IClassFixture<TaxCalculatorFixture<IFederalTaxCalculator>>
    {
        private readonly TaxCalculatorFixture<IFederalTaxCalculator> _fixture;

        public FederalTaxCalculatorTests(TaxCalculatorFixture<IFederalTaxCalculator> fixture)
        {
            _fixture = fixture;
        }

        [Theory(DisplayName = "Federal Tax")]
        [InlineData(2018, 900_000, "Married")]
        [InlineData(2018, 0, "Married")]
        [InlineData(2018, 99995, "Married")]
        public async Task ShouldCalculateFederalTax(int calculationYear, double incomeAsDouble,
            string civilStatusCode)
        {
            // given
            decimal income = Convert.ToDecimal(incomeAsDouble);
            CivilStatus status = Enum.Parse<CivilStatus>(civilStatusCode);

            var taxPerson = new FederalTaxPerson
            {
                Name = "Burli",
                CivilStatus = status,
                TaxableAmount = income
            };

            // when
            var result = await _fixture.Calculator.CalculateAsync(calculationYear, taxPerson);

            result.IsRight.Should().BeTrue();
            Snapshot.Match(result, $"Theory Federal Income Tax {calculationYear}{incomeAsDouble}{civilStatusCode}");

        }
    }
}
