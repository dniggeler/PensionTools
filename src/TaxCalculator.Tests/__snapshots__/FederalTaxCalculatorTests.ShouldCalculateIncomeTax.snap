[
  {
    "State": "IsRight",
    "Right": {
      "DeterminingFactorTaxableAmount": 99900.0,
      "TaxAmount": 1963.0
    },
    "Left": null
  }
]
